### ISSUE TITLE
Error: Push failed for unknown reason.

### ISSUE THREAD

**Issue by MarkPlatts** [workflowr/workflowr#97]
Hi,

When I am trying to push my committed changes to github, I get the above error. Have you any idea why this might be the case?. See below for my diagnostic info:

> sessionInfo()
R version 3.4.3 (2017-11-30)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_0.10.2 rmarkdown_1.8   

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13          getPass_0.2-2         digest_0.6.12         rprojroot_1.3-2      
 [5] R.methodsS3_1.7.1     backports_1.1.1       git2r_0.21.0          magrittr_1.5         
 [9] evaluate_0.10.1       stringi_1.1.6         rstudioapi_0.7.0-9000 R.oo_1.21.0          
[13] R.utils_2.6.0         tools_3.4.3           stringr_1.3.0         yaml_2.1.16          
[17] compiler_3.4.3        htmltools_0.3.6       knitr_1.17        

> Sys.which("git")
                              git 
"C:\\PROGRA~1\\Git\\cmd\\git.exe" 

Any help is much appreciated!

Mark

**Comment by jdblischak** [comment]
Thanks for the diagnostics, @MarkPlatts. This is tricky because `wflow_git_push` is using the R package git2r, which itself is relying on some system libraries to communicate with GitHub.

So your setup looks good:

* Windows 7
* Most recent version of R (3.4.3), workflowr (0.10.2), and git2r (0.21.0)
* Have Git installed

Some followup questions:

* How did you install Git on your Windows machine? Have you run Git commands on your machine prior to installing workflowr? workflowr doesn't actually rely on your system installation of Git, but if we know Git works properly, then git2r should be able to as well.
* What do you get when you run `wflow_remotes()`?

**Comment by MarkPlatts** [comment]
I have had no problems with git on other RStudio projects previously or I
have checked this since the above issue and it has worked fine. It's a
while ago since I installed Git so I'm afraid I can't remember what I did,
other than I think I followed github guidance but I can't be sure.

Heres what I get with wflow_remotes()?

> wflow_remotes()
The repository has the following remotes set:

   name                                                      url
 origin https://github.com/MarkPlatts/Bee_Clustering_Website.git

Could it be something to do with https and the alternative approach. I'm
sorry but I don't have a deep understanding of how git works technically.

Thanks for your help btw.

Mark

**Comment by MarkPlatts** [comment]
I have had no problems with git on other RStudio projects previously *and* I
have checked this since the above issue and it has worked fine

**Comment by MarkPlatts** [comment]
I should also mention that a little while ago I think I had this working
fine. I also seem to remember having this issue and somehow it got
resolved, either of its own accord or I figured something out. Would I get
this error if I used the wrong username and password do you know?

**Comment by pcarbo** [comment]
@MarkPlatts Could you commit a small change to a file in the repo (e.g., a README) and then run `git push origin master` to see if that works for you? If this doesn't work, it is an issue with either the local or remote repo.

**Comment by jdblischak** [comment]
> I have had no problems with git on other RStudio projects previously

@MarkPlatts How do you normally use Git with RStudio projects? Do you use the RStudio Git pane? If yes, are you able to push the workflowr project using the RStudio interface?

> run `git push origin master`

@pcarbo If he didn't install Git via Git Bash, it may not be straightforward to run Git commands in a shell on Windows.

**Comment by MarkPlatts** [comment]
When I try to push from within the RStudio Git pane and give my username and password I get the following which seems strange:

![image](https://user-images.githubusercontent.com/22068402/36874556-446177e6-1da4-11e8-9dc6-d946bd77a904.png)

**Comment by MarkPlatts** [comment]
Using the previous link I have solve the problem preventing me from pushing
from the RStudio Git Pane, but still no joy with workflowr. Still the same
initial error message.

**Comment by MarkPlatts** [comment]
Hi John.

Do you have any other suggestions?

Thanks,
Mark

**Comment by jdblischak** [comment]
Hi Mark. I did a little experimenting on my Windows machine this morning, and also noticed some strange behavior that is likely due to the Issue that you linked above. I need to do more investigation, but I don't have access to my Windows machine currently.

For now, I'd recommend using the RStudio Git pane to push to Github for your workflowr project (from your previous messages it seems that is working for you; please let me know if that is not the case). There is no added benefit of using `wflow_git_push()`. The main purpose of this function is to provide an option for workflowr users that have not installed Git on their machine. Since you've already installed Git and it works, I don't want you to get bogged down trying to troubleshoot this notoriously difficult function. In fact, there was an Issue opened a few days ago that reported similar problems on Windows: https://github.com/ropensci/git2r/issues/329

**Comment by MarkPlatts** [comment]
Ok John. I really appreciate your feedback and support. Thanks!

Mark

**Comment by jdblischak** [comment]
Of course. Happy to help!

Also related:

* https://github.com/ropensci/git2r/issues/327
* https://twitter.com/opencpu/status/966785019560775685
* https://github.com/ropensci/git2r/commit/70cbd3ed0cfc9831ac34c103c2518ed4322dca20

They are working on updating git2r to handle the recent GitHub protocol changes, so I think we'll have to wait until the next release of git2r for `wflow_git_push()` to work reliably on Windows.

**Comment by MarkPlatts** [comment]
Thanks for letting me know.

**Comment by jdblischak** [comment]
@MarkPlatts I finally had the chance to experiment with this. And wow, it's strange behavior. For the most part I had trouble replicating the error. I was able to use `wflow_git_push()` and `wflow_git_pull()` with the HTTPS protocol on Windows 10 with workflowr 0.10.2 and git2r 0.21.0 (the binary version installed from CRAN via `install.packages`, so I didn't compile it against more recent versions of  OpenSSL and libssh2). I tested  an old workflowr repository of mine ([workflowr-demo01](https://github.com/jdblischak/workflowr-demo01)) and a brand new one for testing (not linking here since I'm going to delete it).

I forked your [Bee_Clustering_Website](https://github.com/MarkPlatts/Bee_Clustering_Website) repo (by the way, your workflowr site looks great!), and this was the only time I was able to temporarily reproduce the same error. However, after a few different attempts, it eventually worked, and wouldn't stop working. I tried restarting the R session, restarting my computer, and even deleting and re-forking your repo. Nothing I did was able to bring back the error.

For potential future troubleshooting, here is the code I ran:

```
library("git2r")
packageVersion("git2r") # 0.21.0

tmp <- tempfile("test-windows-")
r <- clone("https://github.com/jdblischak/Bee_Clustering_Website.git",
           local_path = tmp)

library("workflowr") # 0.10.2
wflow_git_push(project = tmp)

credentials <- cred_user_pass("jdblischak", getPass::getPass())
git2r::push(r,
            name = "origin",
            refspec = "refs/heads/master",
            force = FALSE,
            credentials = credentials)
```

If any other Windows users read this thread, my advice is to try `wflow_git_push()` a couple times. It seems that if it works once it will keep working. Hopefully once the new version of git2r is released this behavior will be improved.

In the meantime, I'm going to update workflowr to report the exact error message produced by `git2r::push` and `git2r::pull` to aid future debugging. Unfortunately there's not much more I can do.
